cadquery/occ_impl/shapes.py

```diff
@@ -3984,3 +4094,875 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
+    return _compound_or_shape(builder.Shape())
+
+
+def fill(s: Shape, constraints: Sequence[Union[Shape, VectorLike]] = ()) -> Shape:
+    """
+    Fill edges/wire possibly obeying constraints.
+    """
+
+    builder = BRepOffsetAPI_MakeFilling()
+
+    for e in _get_edges(s):
+        builder.Add(e.wrapped, GeomAbs_C0)
+
+    for c in constraints:
+        if isinstance(c, Shape):
+            for e in _get_edges(c):
+                builder.Add(e.wrapped, GeomAbs_C0, False)
+        else:
+            builder.Add(Vector(c).toPnt())
+
+    builder.Build()
+
+    return _compound_or_shape(builder.Shape())
+
+
+def cap(
+    s: Shape, ctx: Shape, constraints: Sequence[Union[Shape, VectorLike]] = ()
+) -> Shape:
+    """
+    Fill edges/wire possibly obeying constraints and try to connect smoothly to the context shape.
+    """
+
+    builder = BRepOffsetAPI_MakeFilling()
+    builder.SetResolParam(2, 15, 5)
+
+    for e in _get_edges(s):
+        f = _get_one(e.ancestors(ctx, "Face"), "Face")
+        builder.Add(e.wrapped, f.wrapped, GeomAbs_G2, True)
+
+    for c in constraints:
+        if isinstance(c, Shape):
+            for e in _get_edges(c):
+                builder.Add(e.wrapped, GeomAbs_C0, False)
+        else:
+            builder.Add(Vector(c).toPnt())
+
+    builder.Build()
+
+    return _compound_or_shape(builder.Shape())
+
+
+def fillet(s: Shape, e: Shape, r: float) -> Shape:
+    """
+    Fillet selected edges in a given shell or solid.
+    """
+
+    builder = BRepFilletAPI_MakeFillet(_get_one(s, ("Shell", "Solid")).wrapped,)
+
+    for el in _get_edges(e.edges()):
+        builder.Add(r, el.wrapped)
+
+    builder.Build()
+
+    return _compound_or_shape(builder.Shape())
+
+
+def chamfer(s: Shape, e: Shape, d: float) -> Shape:
+    """
+    Chamfer selected edges in a given shell or solid.
+    """
+
+    builder = BRepFilletAPI_MakeChamfer(_get_one(s, ("Shell", "Solid")).wrapped,)
+
+    for el in _get_edges(e.edges()):
+        builder.Add(d, el.wrapped)
+
+    builder.Build()
+
+    return _compound_or_shape(builder.Shape())
+
+
+def extrude(s: Shape, d: VectorLike) -> Shape:
+    """
+    Extrude a shape.
+    """
+
+    results = []
+
+    for el in _get(s, ("Vertex", "Edge", "Wire", "Face")):
+
+        builder = BRepPrimAPI_MakePrism(el.wrapped, Vector(d).wrapped)
+        builder.Build()
+
+        results.append(builder.Shape())
+
+    return _compound_or_shape(results)
+
+
+def revolve(s: Shape, p: VectorLike, d: VectorLike, a: float = 360):
+    """
+    Revolve a shape.
+    """
+
+    results = []
+    ax = gp_Ax1(Vector(p).toPnt(), Vector(d).toDir())
+
+    for el in _get(s, ("Vertex", "Edge", "Wire", "Face")):
+
+        builder = BRepPrimAPI_MakeRevol(el.wrapped, ax, radians(a))
+        builder.Build()
+
+        results.append(builder.Shape())
+
+    return _compound_or_shape(results)
+
+
+def offset(s: Shape, t: float, cap=True, tol: float = 1e-6) -> Shape:
+    """
+    Offset or thicken faces or shells.
```

Comment:
@adam-urbanczyk  Would you consider splitting the offset function into two separate functions `offset` and `thicken`?

The `cap` parameter could then be eliminated. If a thicken function is implemented, what do you think about providing a parameter similar to `Workplane.extrude` `both` to optionally thicken in both directions?

Say if thicken function was available with both=True:
```py
# thicken shell sh0 in both directions
s0 = thicken(sh0, 1, True)
```

It is still possible with current PR to thicken in both directions:
```py
# thicken shell sh0 in both directions
s0 = offset(sh0, 1)
s1 = offset(sh0, -1)
s2 = s0 + s1
```

The thicken function would be comparable to the thicken Face method except that it also supports Shell.
```py
e0 = segment((0, 0), (10, 10))
f0 = extrude(e0, (0, 0, 10))
s0 = f0.thicken(1)   # existing Face method
s1 = thicken(f0, 1)  # this PR
```

I'm not suggesting it is a Must fix/change if you are against it and already rejected the idea.

Reply:
Currently `offset` has the same behavior as `Face.thicken`, i.e. it is not symmetric. I don't want to add more functionality to this PR, but the API will for sure change - it should be for now considered experimental. `thicken` free func could be added in future PRs.